File: app/src/main/java/de/txserver/slickupnp/activity/MainActivity.java
package de.txserver.slickupnp.activity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.widget.DrawerLayout;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;

import de.txserver.slickupnp.helper.DrawerListAdapter;
import de.txserver.slickupnp.helper.NavigationItem;
import de.txserver.slickupnp.upnp.ContentDirectoryBrowseCallbacks;
import de.txserver.slickupnp.R;
import de.txserver.slickupnp.helper.CustomListAdapter;
import de.txserver.slickupnp.helper.CustomListItem;
import de.txserver.slickupnp.helper.DeviceModel;
import de.txserver.slickupnp.helper.ItemModel;
import de.txserver.slickupnp.upnp.ContentDirectoryBrowseHandler;

public class MainActivity extends AppCompatActivity
                          implements ContentDirectoryBrowseCallbacks,
                          SharedPreferences.OnSharedPreferenceChangeListener {

    public String TAG = MainActivity.class.getSimpleName();

    private ActionBar actionBar;
    private ArrayList<NavigationItem> navigationItemArrayList;

    private DrawerLayout drawerLayout;
    private RelativeLayout drawerPane;
    private ActionBarDrawerToggle drawerToggle;
    private SwipeRefreshLayout swipeRefreshLayout;

    private ListView listView;

    private ContentDirectoryBrowseHandler contentDirectoryBrowseHandler;
    private ArrayList<CustomListItem> mDeviceList;
    private ArrayAdapter<CustomListItem> mDeviceListAdapter;
    private ArrayList<CustomListItem> mItemList;
    private ArrayAdapter<CustomListItem> mItemListAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        Toolbar myToolbar = (Toolbar) findViewById(R.id.myToolbar);
        setSupportActionBar(myToolbar);

        actionBar = getSupportActionBar();
        if (actionBar != null) {

            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }

        navigationItemArrayList = new ArrayList<>();
        navigationItemArrayList.add(new NavigationItem(getResources().getString(R.string.menuItem_settings), getResources().getString(R.string.menuItemText_settings), R.drawable.ic_settings_black_24px, R.layout.activity_settings));
        navigationItemArrayList.add(new NavigationItem(getResources().getString(R.string.menuItem_about), getResources().getString(R.string.menuItemText_about), R.drawable.ic_info_outline_black_24px, R.layout.activity_about));

        // Populate the Navigtion Drawer with options
        drawerLayout = (DrawerLayout)findViewById(R.id.drawerLayout);
        drawerPane = (RelativeLayout)findViewById(R.id.drawerPane);

        ListView navigationListView = (ListView)findViewById(R.id.navList);
        DrawerListAdapter drawerListAdapter = new DrawerListAdapter(this, navigationItemArrayList);
        navigationListView.setAdapter(drawerListAdapter);
        navigationListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                selectItemFromDrawer(position);
            }
        });

        drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, R.string.drawer_open, R.string.drawer_close) {

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                invalidateOptionsMenu();
            }

            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
                invalidateOptionsMenu();
            }
        };

        drawerLayout.addDrawerListener(drawerToggle);

        swipeRefreshLayout = (SwipeRefreshLayout)findViewById(R.id.swipeRefresh);
        swipeRefreshLayout.setOnRefreshListener(
                new SwipeRefreshLayout.OnRefreshListener() {
                    @Override
                    public void onRefresh() {
                        refreshList();
                    }
                }
        );


        mDeviceList = new ArrayList<CustomListItem>();
        mDeviceListAdapter = new CustomListAdapter(this, mDeviceList);

        mItemList = new ArrayList<CustomListItem>();
        mItemListAdapter = new CustomListAdapter(this, mItemList);

        listView = (ListView)findViewById(R.id.list);
        listView.setAdapter(mDeviceListAdapter);

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {

            @Override
            public void onItemClick (AdapterView < ? > adapter, View v,int position, long arg3){
                setShowRefreshing(true);
                contentDirectoryBrowseHandler.navigateTo(adapter.getItemAtPosition(position));
            }
        });

        if (contentDirectoryBrowseHandler == null) {
            contentDirectoryBrowseHandler = new ContentDirectoryBrowseHandler(this);
            contentDirectoryBrowseHandler.bindServiceConnection();
        } else {
            contentDirectoryBrowseHandler.refreshDevices();
            contentDirectoryBrowseHandler.refreshCurrent();
        }

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        prefs.registerOnSharedPreferenceChangeListener(this);

        final IntentFilter filter = new IntentFilter();
        filter.addAction("android.net.wifi.WIFI_STATE_CHANGED");
        registerReceiver(receiver, filter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.action_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (drawerToggle.onOptionsItemSelected(item)) {
            return true;
        } else {
            switch (item.getItemId()) {
                case R.id.action_refresh:
//                    Toast.makeText(this, R.string.info_searching, Toast.LENGTH_SHORT).show();
                    refreshList();
                    break;
            }
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {

        super.onPostCreate(savedInstanceState);

        drawerToggle.syncState();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        unregisterReceiver(receiver);
        contentDirectoryBrowseHandler.unbindServiceConnection();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences prefs, String key) {
        refreshList();
    }

    private void setActivityTitle() {

        if (actionBar == null) {
            return;
        }

        actionBar.setTitle(getResources().getString(R.string.app_name));
    }

    private void selectItemFromDrawer(int position) {

        Intent intent = null;

        switch(navigationItemArrayList.get(position).getActivity()){

            case R.layout.activity_settings:
                intent = new Intent(this, SettingsActivity.class);
                break;
            case R.layout.activity_about:
                intent = new Intent(this, AboutActivity.class);
                break;
        }

        if (intent != null) {

            startActivity(intent);
        }

        // Close the drawer
        drawerLayout.closeDrawer(drawerPane);
    }

    private void unsetListSelection() {
        listView.setItemChecked(-1, true);
    }

    private void refreshList() {
        setShowRefreshing(true);
        contentDirectoryBrowseHandler.refreshCurrent();
    }

    @Override
    public void setShowRefreshing(final boolean show) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                swipeRefreshLayout.setRefreshing(show);
            }
        });
    }

    @Override
    public void onBackPressed() {
        if (contentDirectoryBrowseHandler.goBack())
            super.onBackPressed();
    }

    @Override
    public void onDisplayDevices() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                unsetListSelection();
                mDeviceListAdapter.clear();
                listView.setAdapter(mDeviceListAdapter);
                setShowRefreshing(false);
            }
        });
    }

    @Override
    public void onDisplayDirectories() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                unsetListSelection();
                mItemList.clear();
                mItemListAdapter.clear();
                listView.setAdapter(mItemListAdapter);
                setShowRefreshing(false);
            }
        });
    }

    @Override
    public void onDisplayItems(final ArrayList<ItemModel> items) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                unsetListSelection();
                mItemList.clear();
                mItemList.addAll(items);
                mItemListAdapter.notifyDataSetChanged();
            }
        });
    }

    @Override
    public void onDisplayAddItems(final ArrayList<ItemModel> items) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mItemList.addAll(items);
                mItemListAdapter.notifyDataSetChanged();
            }
        });
    }

    @Override
    public void onDisplayItemsError(final String error) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mItemListAdapter.clear();
                mItemListAdapter.add(new CustomListItem(
                        R.drawable.ic_alert_black_24px,
                        getResources().getString(R.string.info_errorlist_folders),
                        error));
            }
        });
    }

    @Override
    public void onDeviceAdded(final DeviceModel device) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                int position = mDeviceListAdapter.getPosition(device);
                if (position >= 0) {
                    mDeviceListAdapter.remove(device);
                    mDeviceListAdapter.insert(device, position);
                } else {
                    mDeviceListAdapter.add(device);
                }
            }
        });
    }

    @Override
    public void onDeviceRemoved(final DeviceModel device) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mDeviceListAdapter.remove(device);
            }
        });
    }

    private BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String action = intent.getAction();
            if (action.equals(WifiManager.WIFI_STATE_CHANGED_ACTION)) {

                int state = intent.getIntExtra(WifiManager.EXTRA_WIFI_STATE,
                        WifiManager.WIFI_STATE_UNKNOWN);

                TextView wifi_warning = (TextView)findViewById(R.id.wifi_warning);

                switch (state) {
                    case WifiManager.WIFI_STATE_ENABLED:
                        wifi_warning.setVisibility(View.GONE);

                        if (contentDirectoryBrowseHandler != null) {
                            contentDirectoryBrowseHandler.refreshDevices();
                            contentDirectoryBrowseHandler.refreshCurrent();
                        }
                        break;
                    case WifiManager.WIFI_STATE_DISABLED:
                        wifi_warning.setVisibility(View.VISIBLE);
                        mDeviceListAdapter.clear();
                        mItemListAdapter.clear();
                        break;
                    case WifiManager.WIFI_STATE_UNKNOWN:
                        wifi_warning.setVisibility(View.VISIBLE);
                        break;
                }
            }
        }
    };
}

File: app/src/main/java/de/txserver/slickupnp/activity/MainActivity.kt
package de.txserver.slickupnp.activity

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.content.res.Resources
import android.net.wifi.WifiManager
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.content.ContextCompat
import android.support.v4.widget.DrawerLayout
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.ActionBar
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.RelativeLayout
import android.widget.TextView

import java.util.ArrayList

import de.txserver.slickupnp.helper.DrawerListAdapter
import de.txserver.slickupnp.helper.NavigationItem
import de.txserver.slickupnp.upnp.ContentDirectoryBrowseCallbacks
import de.txserver.slickupnp.R
import de.txserver.slickupnp.helper.CustomListAdapter
import de.txserver.slickupnp.helper.CustomListItem
import de.txserver.slickupnp.helper.DeviceModel
import de.txserver.slickupnp.helper.ItemModel
import de.txserver.slickupnp.upnp.ContentDirectoryBrowseHandler

class MainActivity : AppCompatActivity(), ContentDirectoryBrowseCallbacks, SharedPreferences.OnSharedPreferenceChangeListener {

    private val TAG = MainActivity::class.java.simpleName;

    private var actionBar: ActionBar? = null;
    private var navigationItemArrayList: ArrayList<NavigationItem>? = null;

    private var drawerLayout: DrawerLayout? = null
    private var drawerPane: RelativeLayout? = null
    private var drawerToggle: ActionBarDrawerToggle? = null
    private var swipeRefreshLayout: SwipeRefreshLayout? = null

    private var listView: ListView? = null

    private var contentDirectoryBrowseHandler: ContentDirectoryBrowseHandler? = null
    private var mDeviceList: ArrayList<CustomListItem>? = null
    private var mDeviceListAdapter: ArrayAdapter<CustomListItem>? = null
    private var mItemList: ArrayList<CustomListItem>? = null
    private var mItemListAdapter: ArrayAdapter<CustomListItem>? = null

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (action == WifiManager.WIFI_STATE_CHANGED_ACTION) {

                val state = intent.getIntExtra(WifiManager.EXTRA_WIFI_STATE,
                        WifiManager.WIFI_STATE_UNKNOWN)

                val wifi_warning = findViewById(R.id.wifi_warning) as TextView

                when (state) {
                    WifiManager.WIFI_STATE_ENABLED -> {
                        wifi_warning.visibility = View.GONE

                        contentDirectoryBrowseHandler?.let {
                            it.refreshDevices()
                            it.refreshCurrent()
                        }
                    }
                    WifiManager.WIFI_STATE_DISABLED -> {
                        wifi_warning.visibility = View.VISIBLE
                        mDeviceListAdapter?.clear()
                        mItemListAdapter?.clear()
                    }
                    WifiManager.WIFI_STATE_UNKNOWN -> wifi_warning.visibility = View.VISIBLE
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        val myToolbar = findViewById(R.id.myToolbar) as Toolbar
        setSupportActionBar(myToolbar)

        actionBar = supportActionBar

        actionBar?.let {
            it.setDisplayHomeAsUpEnabled(true)
            it.setHomeButtonEnabled(true)
        }

        navigationItemArrayList = ArrayList()

        navigationItemArrayList?.let {
            it.add(NavigationItem(resources.getString(R.string.menuItem_settings), resources.getString(R.string.menuItemText_settings), R.drawable.ic_settings_black_24px, R.layout.activity_settings))
            it.add(NavigationItem(resources.getString(R.string.menuItem_about), resources.getString(R.string.menuItemText_about), R.drawable.ic_info_outline_black_24px, R.layout.activity_about))
            it.add(NavigationItem(resources.getString(R.string.menuItem_exit), resources.getString(R.string.menuItemText_exit), R.drawable.ic_exit_to_app_black_24px, R.layout.activity_main))
        }

        // Populate the Navigtion Drawer with options
        drawerLayout = findViewById(R.id.drawerLayout) as DrawerLayout
        drawerPane = findViewById(R.id.drawerPane) as RelativeLayout

        val navigationListView = findViewById(R.id.navList) as ListView
        val drawerListAdapter = DrawerListAdapter(this, navigationItemArrayList)
        navigationListView.adapter = drawerListAdapter
        navigationListView.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id -> selectItemFromDrawer(position) }

        drawerToggle = object : ActionBarDrawerToggle(this, drawerLayout, R.string.drawer_open, R.string.drawer_close) {

            override fun onDrawerOpened(drawerView: View?) {
                super.onDrawerOpened(drawerView)
                invalidateOptionsMenu()
            }

            override fun onDrawerClosed(drawerView: View?) {
                super.onDrawerClosed(drawerView)
                invalidateOptionsMenu()
            }
        }

        drawerLayout?.addDrawerListener(drawerToggle!!)

        swipeRefreshLayout = findViewById(R.id.swipeRefresh) as SwipeRefreshLayout

        swipeRefreshLayout?.let {
            it.setColorSchemeColors(ContextCompat.getColor(this, R.color.colorPrimary))
            it.setOnRefreshListener { refreshList() }
        }


        mDeviceList = ArrayList()
        mDeviceListAdapter = CustomListAdapter(this, mDeviceList)

        mItemList = ArrayList()
        mItemListAdapter = CustomListAdapter(this, mItemList)

        listView = findViewById(R.id.list) as ListView
        listView?.adapter = mDeviceListAdapter

        listView?.onItemClickListener = AdapterView.OnItemClickListener { adapter, v, position, arg3 ->
            setShowRefreshing(true)
            contentDirectoryBrowseHandler?.navigateTo(adapter.getItemAtPosition(position))
        }

        if (contentDirectoryBrowseHandler != null) {
            contentDirectoryBrowseHandler?.refreshDevices()
            contentDirectoryBrowseHandler?.refreshCurrent()
        } else {
            contentDirectoryBrowseHandler = ContentDirectoryBrowseHandler(this)
            contentDirectoryBrowseHandler?.bindServiceConnection()
        }

        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        prefs.registerOnSharedPreferenceChangeListener(this)

        val filter = IntentFilter()
        filter.addAction("android.net.wifi.WIFI_STATE_CHANGED")
        registerReceiver(receiver, filter)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.action_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (drawerToggle!!.onOptionsItemSelected(item)) {
            return true
        } else {
            when (item.itemId) {
                R.id.action_refresh -> refreshList()
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {

        super.onPostCreate(savedInstanceState)

        drawerToggle?.syncState()
    }

    public override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(receiver)
        contentDirectoryBrowseHandler?.unbindServiceConnection()
    }

    override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String) {
        refreshList()
    }

    private fun setActivityTitle() {

        if (actionBar == null) {
            return
        }

        actionBar?.setTitle(resources.getString(R.string.app_name))
    }

    private fun selectItemFromDrawer(position: Int) {

        var intent: Intent? = null

        when (navigationItemArrayList!![position].activity) {

            R.layout.activity_settings -> intent = Intent(this, SettingsActivity::class.java)
            R.layout.activity_about -> intent = Intent(this, AboutActivity::class.java)
            R.layout.activity_main -> this.finish()
        }

        if (intent != null) {

            startActivity(intent)
        }

        // Close the drawer
        drawerLayout?.closeDrawer(drawerPane)
    }

    private fun unsetListSelection() {
        listView?.setItemChecked(-1, true)
    }

    private fun refreshList() {
        setShowRefreshing(true)
        contentDirectoryBrowseHandler?.refreshCurrent()
    }

    override fun setShowRefreshing(show: Boolean) {
        runOnUiThread { swipeRefreshLayout?.isRefreshing = show }
    }

    override fun onBackPressed() {
        if (contentDirectoryBrowseHandler?.goBack()!!)
            super.onBackPressed()
    }

    override fun onDisplayDevices() {
        runOnUiThread {
            unsetListSelection()
            mDeviceListAdapter?.clear()
            listView?.adapter = mDeviceListAdapter
            setShowRefreshing(false)
        }
    }

    override fun onDisplayDirectories() {
        runOnUiThread {
            unsetListSelection()
            mItemList?.clear()
            mItemListAdapter?.clear()
            listView?.adapter = mItemListAdapter
            setShowRefreshing(false)
        }
    }

    override fun onDisplayItems(items: ArrayList<ItemModel>) {
        runOnUiThread {
            unsetListSelection()
            mItemList?.clear()
            mItemList?.addAll(items)
            mItemListAdapter?.notifyDataSetChanged()
        }
    }

    override fun onDisplayAddItems(items: ArrayList<ItemModel>) {
        runOnUiThread {
            mItemList?.addAll(items)
            mItemListAdapter?.notifyDataSetChanged()
        }
    }

    override fun onDisplayItemsError(error: String) {
        runOnUiThread {
            mItemListAdapter?.clear()
            mItemListAdapter?.add(CustomListItem(
                    R.drawable.ic_alert_black_24px,
                    resources.getString(R.string.info_errorlist_folders),
                    error))
        }
    }

    override fun onDeviceAdded(device: DeviceModel) {
        runOnUiThread {
            val position = mDeviceListAdapter?.getPosition(device) ?: 0
            if (position >= 0) {
                mDeviceListAdapter?.remove(device)
                mDeviceListAdapter?.insert(device, position)
            } else {
                mDeviceListAdapter?.add(device)
            }
        }
    }

    override fun onDeviceRemoved(device: DeviceModel) {
        runOnUiThread { mDeviceListAdapter?.remove(device) }
    }
}